### PR TITLE
Add macOS and tvOS push notification support

### PR DESCRIPTION
**Pull request by funkenstrahlen**
Make push notifications work for macOS and tvOS.

This is linked to this issue: https://github.com/parse-server-modules/parse-server-push-adapter/issues/11

- [x] Add support for osx and tvos deviceTypes to installation
- [x] Add support to ParsePushAdapter and APNS.js to accept more than just iOS certificates.
- [x] Add support to ParsePushAdapter and APNS.js to send to tvOS, which reuses the same certificate as iOS.
- [x] Update spec

### CODE REVIEW

**Comment by funkenstrahlen** [comment]
@flovilmart This is what I came up with. Can you please take a look?

File: src/ParsePushAdapter.js [hunk 1]
```diff
@@ -49,8 +51,19 @@ export class ParsePushAdapter {
     let deviceMap = classifyInstallations(installations, this.validPushTypes);
     let sendPromises = [];
     for (let pushType in deviceMap) {
-      let sender = this.senderMap[pushType];
+      var sender = this.senderMap[pushType];
       let devices = deviceMap[pushType];
+      
+      if (!sender) {
+        if (pushType == 'tvos') {
```
Comment (flovilmart) [inline src/ParsePushAdapter.js]: not sure this fallback is required. Even if it leads to redundancy for all 'apple' push types.

**Comment by funkenstrahlen** [comment]
Is `tvos` actually the correct deviceType for the Apple TV? Or do I have to use `appletv`?

**Comment by funkenstrahlen** [comment]
I need some help on the last spec test. I do not know why it fails with cannot found valid connection.

I added some debug logging and saw this:

```
.ERR! parse-server-push-adapter calling send for devices [object Object],[object Object]
ERR! parse-server-push-adapter APNS found conn iosbundleId for: 0d72a1baa92a2febd9a254cbd6584f750c70b2350af5fc9052d1d12584b738e6
ERR! parse-server-push-adapter APNS found conn iosbundleId for: ff3943ed0b2090c47e5d6f07d8f202a10427941d7897fda5a6b18c6d9fd07d48
ERR! parse-server-push-adapter calling send for devices [object Object]
ERR! parse-server-push-adapter APNS found conn osxbundleId for: 5cda62a8d88eb48d9111a6c436f2e326a053eb0cd72dfc3a0893089342602235
ERR! parse-server-push-adapter calling send for devices [object Object]
ERR! parse-server-push-adapter APNS found conn iosbundleId for: 3e72a1baa92a2febd9a254cbd6584f750c70b2350af5fc9052d1d12584b738e6
ERR! parse-server-push-adapter calling send for devices [object Object]
ERR! parse-server-push-adapter APNS cannot find vaild connection for ff3943ed0b2090c47e5d6f07d8f202a10427941d7897fda5a6b18c6d9fd07d48
ERR! parse-server-push-adapter APNS cannot find vaild connection for 0d72a1baa92a2febd9a254cbd6584f750c70b2350af5fc9052d1d12584b738e6
ERR! parse-server-push-adapter APNS cannot find vaild connection for 3e72a1baa92a2febd9a254cbd6584f750c70b2350af5fc9052d1d12584b738e6
```

It groups devices correctly: `ios`, `osx`, `android`, `tvos`. It also find matching connection configuration for each group and uses the `ios` configurated connection for `tvos` deviceType.

However it reports invalid connection for ios and tvos devices later. Any idea what part in the code causes this problem?

**Comment by funkenstrahlen** [comment]
I spend hours trying to figure out why this test case does fail. Because I did not know what change caused the test case to fail I switched back to the master branch. Result: The test case does not run either!

```
npm test

> parse-server-push-adapter@1.2.0 test /Users/funkenstrahlen/Projects/podlive/parse-server-push-adapter
> TESTING=1 babel-node ./node_modules/.bin/isparta cover --root src/ ./node_modules/jasmine/bin/jasmine.js

Started
...................ERR! parse-server-push-adapter GCM send errored: 401
........ERR! parse-server-push-adapter APNS no qualified connections for anotherBundle c5ee8fae0a1c
ERR! parse-server-push-adapter APNS no qualified connections for anotherBundle c5ee8fae0a1c5805e731cf15496d5b2b3f9b9c577353d3239429d3faaee01c79
ERR! parse-server-push-adapter GCM send errored: 401
.


28 specs, 0 failures
Finished in 0.448 seconds

No coverage information was collected, exit without writing coverage information
```

So I do not see how I can make this test pass if it does not even work in your current master.

**Comment by flovilmart** [comment]
The print out shows that the tests are passing. (0 failures)

**Comment by funkenstrahlen** [comment]
So this is ready to merge?

**Comment by flovilmart** [comment]
I'll check it out on my machine and see what can fail

**Comment by flovilmart** [comment]
I found the reason of the failures (APNs is throwing an exception when the certificate is missing), and I'Ll try to provide a fix on the top of your branch, Is that ok for you?

**Comment by funkenstrahlen** [comment]
> I found the reason of the failures (APNs is throwing an exception when the certificate is missing), and I'Ll try to provide a fix on the top of your branch, Is that ok for you?

This is great news! Of course you can add your changes. I granted you push access to my fork in case it is required.

**Comment by flovilmart** [comment]
> I granted you push access to my fork in case it is required.

That's not required, I can push on the PR branch directly :) Thanks!

**Comment by flovilmart** [comment]
@funkenstrahlen I uncovered additional inconsistencies with our implementation. I'm still working on it :)

File: src/ParsePushAdapter.js (hunk 1 above)
Comment (funkenstrahlen) [inline src/ParsePushAdapter.js]: I added this as it was suggested in the related issue. However I also think it is better to not add redundancy and require the admin to define a certificate for each plattform he wants to support.

**Comment by funkenstrahlen** [comment]
It is 'tvos' or 'appletv'?

File: src/ParsePushAdapter.js [hunk 2]
```diff
@@ -49,8 +51,19 @@ export class ParsePushAdapter {
...
+          // if there is only an ios push configuration available reuse it for tvos
+          sender = this.senderMap['ios'];
+        } else if (pushType == 'ios') {
+          // if there is only a tvos push configuration available reuse it for ios
+          sender = this.senderMap['tvos'];
```
Comment (funkenstrahlen) [inline src/ParsePushAdapter.js]: Are you suggesting to also add `macos` into this redundancy?

**Comment by funkenstrahlen** [comment]
> I uncovered additional inconsistencies with our implementation. I'm still working on it :)

Thanks for fixing the spec! Anything further I can do to help getting this merged?

File: src/ParsePushAdapter.js (hunk 2 above)
Comment (flovilmart) [inline src/ParsePushAdapter.js]: Actually I'm not sure if we should automatically do that. Do you Think of any real world use case?

Comment (funkenstrahlen) [inline src/ParsePushAdapter.js]: I personally think is is more transparent if we do not add this redundancy. If someone needs support for multiple push targets he needs to define all of them in his config. He can use the same certificate of course.

Comment (funkenstrahlen) [inline src/ParsePushAdapter.js]: If you agree on this I will remove this redundancy code and adjust the spec.

Comment (flovilmart) [inline src/ParsePushAdapter.js]: Fully agreed! less code, better code!

**Comment by funkenstrahlen** [comment]
@flovilmart I removed the redundancy push configuration code as we discussed and also adjusted the spec.

**Comment by flovilmart** [comment]
Yeah! Took some commits but we're there!!

**Comment by funkenstrahlen** [comment]
Awesome! Thanks for helping me getting this done! :)

Do you think this is worth a version bump?

**Comment by flovilmart** [comment]
Yep definitely! We should probably release it :)

**Pull request merged**